### ISSUE TITLE
Stream handler channelRead method invoked before channelActive

### ISSUE THREAD

**Issue by mbaxter** [libp2p/jvm-libp2p#94]
### Issue Description
I've been testing teku's syncing logic and have noticed an issue with the ordering of events for stream handlers.  For reference, our stream-related setup logic is [here](https://github.com/PegaSysEng/artemis/blob/c5748972645aaf4ef937e99419e06c957398d06a/networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java#L39).

When a remote peer initiates a request stream with our node, we push a `SimpleChannelInboundHandler` to manage these requests and I've noticed that this handler's `channelRead0` method is sometimes invoked before `channelActive`.  From discussing with @Nashatyrev, it sounds like this may be a bug.

**Comment by mbaxter** [comment]
I added a bit more logging on a branch here: https://github.com/mbaxter/artemis/tree/debug/libp2p-incoming-rpc-req-error and captured some logs related to this issue.

I've included 4000 logs surrounding the error `2020-02-13 15:05:24.906-05:00 | nioEventLoopGroup-3-10 | ERROR | Eth2IncomingRequestHandler | Unhandled error while processing request from 16Uiu2HAkyz91Gb8F6pPpD5tDrFGnce34ozy8KSzXkWHeqKi3bAek  :/eth2/beacon_chain/req/status/1/ssz` here: [jvm-libp2p-issue-93.log](https://github.com/libp2p/jvm-libp2p/files/4200814/jvm-libp2p-issue-93.log)

And here I've filtered all of the logs looking for anything related to the libp2p handling (`cat artemis.log | egrep -i "194.55.12.148|id: 0xb8e8e46f|Ctrl 140286422|Unhandled error while processing request"`) here: 
[jvm-libp2p-issue-93-wire.log](https://github.com/libp2p/jvm-libp2p/files/4200907/jvm-libp2p-issue-93-wire.log)

These errors relate to the peer `ip4/194.55.12.148/tcp/13000/p2p/16Uiu2HAkyz91Gb8F6pPpD5tDrFGnce34ozy8KSzXkWHeqKi3bAek`

The error appears to be a result of the fact that [readChannel0](https://github.com/mbaxter/artemis/blob/507d36c01566ea53e9e9645bfed5d7b01dbda68f/networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java#L124) is called before [channelActive](https://github.com/mbaxter/artemis/blob/507d36c01566ea53e9e9645bfed5d7b01dbda68f/networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java#L112), which means that `rpcStream` is `null`.

**Comment by Nashatyrev** [comment]
```
2020-02-13 15:05:24.905-05:00 | nioEventLoopGroup-3-10 | DEBUG | mux | [id: 0xb8e8e46f, L:/192.168.1.131:51513 - R:/194.55.12.148:13000] READ: MplexFrame(id=MuxId(id=4, initiator=false), flag=DATA (init), data=132f6d756c746973747265616d2f312e302e300a242f657468322f626561636f6e5f636861696e2f7265712f7374617475732f312f73737a0a540000000415865fa27f526792236aacdd44ce6deb068c591cc6909a1883bb15529c40120a981e00000000000065888b41ac6b0729dbc1692b13d6d3610f31be2f7280db9c82376c4aa1f1807fc3d3030000000000), 142B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 13 2f 6d 75 6c 74 69 73 74 72 65 61 6d 2f 31 2e |./multistream/1.|
|00000010| 30 2e 30 0a 24 2f 65 74 68 32 2f 62 65 61 63 6f |0.0.$/eth2/beaco|
|00000020| 6e 5f 63 68 61 69 6e 2f 72 65 71 2f 73 74 61 74 |n_chain/req/stat|
|00000030| 75 73 2f 31 2f 73 73 7a 0a 54 00 00 00 04 15 86 |us/1/ssz.T......|
|00000040| 5f a2 7f 52 67 92 23 6a ac dd 44 ce 6d eb 06 8c |_..Rg.#j..D.m...|
|00000050| 59 1c c6 90 9a 18 83 bb 15 52 9c 40 12 0a 98 1e |Y........R.@....|
|00000060| 00 00 00 00 00 00 65 88 8b 41 ac 6b 07 29 db c1 |......e..A.k.)..|
|00000070| 69 2b 13 d6 d3 61 0f 31 be 2f 72 80 db 9c 82 37 |i+...a.1./r....7|
|00000080| 6c 4a a1 f1 80 7f c3 d3 03 00 00 00 00 00       |lJ............  |
+--------+-------------------------------------------------+----------------+
```
That packet looks suspicious in the sense it contains both `multiprotocol` negotiation last packet and the client protocol request. So if any asynchronous code is involved in calling `channelActive` before `channelRead` then it may be the key for addressing this issue.

**Comment by Nashatyrev** [comment]
> That packet looks suspicious

We are not capable now of processing this kind of '0-roundtrip' `multistream` negotiation: i.e. the multistream protocol name if followed by the payload in a single packet. 
Though it is absolutely good idea I'm not 100% confident that such behavior is spec compatible as the initiator sends request data without making sure that responder supports the requested protocol.

That behavior could be adjusted by fixing this `multistream` negotiator sequence: 
https://github.com/libp2p/jvm-libp2p/blob/e7695b9ac99af04bc6184dc0d6cf7f06eb8b1197/src/main/kotlin/io/libp2p/multistream/Negotiator.kt#L99-L103

Currently the sequence of events is incorrect since while removing `ProtobufVarint32FrameDecoder` it flushes buffered frame which synchronously propagated to the protocol handler without calling `channelActive` first (which is fired a bit later after removing handlers).
Though we couldn't simply call `ctx.fireChannelActive()` before removing as the protocol handler may want to send some data on activation, but there are still unremoved obsolete handlers on the pipeline

**Comment by Nashatyrev** [comment]
Should be fixed with `0.5.1`
